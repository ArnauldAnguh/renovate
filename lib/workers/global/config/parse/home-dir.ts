import os from 'os';
import fs from 'fs-extra';
import ini from 'ini';
import { join } from 'upath';
import { AllConfig } from '../../../../config/types';
import { logger } from '../../../../logger';

export interface HomeDirConfig {
  fileList: string[];
  config: AllConfig;
}

export async function readHomeDirFiles(): Promise<HomeDirConfig | null> {
  const fileList: string[] = [];
  const config: AllConfig = { hostRules: [], packageRules: [] };
  const homeDir = os.homedir();
  const npmrcFileName = join(homeDir, '.npmrc');
  let npmrcContent: string;
  try {
    npmrcContent = await fs.readFile(npmrcFileName, 'utf8');
  } catch (err) {
    // Ignore
  }
  if (npmrcContent) {
    fileList.push(npmrcFileName);
    try {
      const npmrc = ini.parse(npmrcContent);
      if (npmrc.registry) {
        config.packageRules.push({
          matchDatasources: ['npm'],
          registryUrls: [npmrc.registry],
        });
      }
      if (npmrc._authToken) {
        config.hostRules.push({ hostType: 'npm', token: npmrc._authToken });
      }
      if (npmrc._auth) {
        config.hostRules.push({
          hostType: 'npm',
          token: npmrc._auth,
          authType: 'Basic',
        });
      }
      for (const [key, val] of Object.entries(npmrc)) {
        const [scope, topic] = key.split(':');
        if (topic === 'registry') {
          config.packageRules.push({
            matchDatasources: ['npm'],
            matchPackagePrefixes: [scope],
            registryUrls: [val],
          });
        }
        if (topic === '_authToken') {
          const matchHost = scope.replace(/^\/\//, 'https://');
          config.hostRules.push({ matchHost, token: val });
        }
        if (topic === '_auth') {
          const matchHost = scope.replace(/^\/\//, 'https://');
          config.hostRules.push({
            matchHost,
            token: val,
            authType: 'Basic',
          });
        }
      }
    } catch (err) /* istanbul ignore next */ {
      logger.debug({ npmrcFileName, err }, 'Error parsing .npmrc');
    }
  }
  if (fileList.length) {
    return { fileList, config };
  }
  return null;
}
