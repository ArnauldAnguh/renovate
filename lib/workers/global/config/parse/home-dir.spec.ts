import * as _fs from 'fs-extra';
import { getName, mocked } from '../../../../../test/util';
import { readHomeDirFiles } from './home-dir';

jest.mock('fs-extra');

const fs = mocked(_fs);

describe(getName(), () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });
  describe('readHomeDirFiles()', () => {
    it('returns null if no file', async () => {
      expect(await readHomeDirFiles()).toBeNull();
    });
    it('parses .npmrc', async () => {
      const content = [
        'registry=https://registry.npmjs.org/',
        '_auth=abc123',
        '_authToken=def456',
        '//registry.npmjs.org/:_authToken=aaaaaa111111-bbbbbb222222',
        '//registry.npmjs.org/:_auth=afafafaf==',
        '@renovate:registry=https://registry.renovatebot.com/',
      ];
      fs.readFile.mockResolvedValueOnce(content.join('\n') as any);
      expect((await readHomeDirFiles()).config).toMatchSnapshot();
    });
  });
});
