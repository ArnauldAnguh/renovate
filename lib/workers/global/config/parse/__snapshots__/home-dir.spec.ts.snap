// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workers/global/config/parse/home-dir readHomeDirFiles() parses .npmrc 1`] = `
Object {
  "hostRules": Array [
    Object {
      "hostType": "npm",
      "token": "def456",
    },
    Object {
      "authType": "Basic",
      "hostType": "npm",
      "token": "abc123",
    },
    Object {
      "matchHost": "https://registry.npmjs.org/",
      "token": "aaaaaa111111-bbbbbb222222",
    },
    Object {
      "authType": "Basic",
      "matchHost": "https://registry.npmjs.org/",
      "token": "afafafaf==",
    },
  ],
  "packageRules": Array [
    Object {
      "matchDatasources": Array [
        "npm",
      ],
      "registryUrls": Array [
        "https://registry.npmjs.org/",
      ],
    },
    Object {
      "matchDatasources": Array [
        "npm",
      ],
      "matchPackagePrefixes": Array [
        "@renovate",
      ],
      "registryUrls": Array [
        "https://registry.renovatebot.com/",
      ],
    },
  ],
}
`;
